openedx/core/djangoapps/content/course_structures/api/v0/api.py

```diff
@@ -23,52 +24,75 @@ def _retrieve_course(course_key):
 
     """
-    try:
-        return courses.get_course(course_key)
-    except ValueError:
+    course = modulestore().get_course(course_key, depth=0)
+    if course is None:
         raise CourseNotFoundError
 
+    return course
 
-def course_structure(course_key):
-    """
-    Retrieves the entire course structure, including information about all the blocks used in the course.
 
+def course_structure(course_key, block_types=None):
+    """
+    Retrieves the entire course structure, including information about all the blocks used in the
+    course if `block_types` is None else information about `block_types` will be returned only.
+    Final serialized information will be cached.
     Args:
         course_key: the CourseKey of the course we'd like to retrieve.
+        block_types: list of required block types. Possible values include sequential,
+                     vertical, html, problem, video, and discussion. The type can also be
+                     the name of a custom type of block used for the course.
     Returns:
         The serialized output of the course structure:
             * root: The ID of the root node of the course structure.
-
             * blocks: A dictionary that maps block IDs to a collection of
             information about each block. Each block contains the following
             fields.
-
                 * id: The ID of the block.
-
                 * type: The type of block. Possible values include sequential,
                     vertical, html, problem, video, and discussion. The type can also be
                     the name of a custom type of block used for the course.
-
                 * display_name: The display name configured for the block.
-
                 * graded: Whether or not the sequential or problem is graded. The
                     value is true or false.
-
                 * format: The assignment type.
-
                 * children: If the block has child blocks, a list of IDs of the child
                 blocks.
     Raises:
         CourseStructureNotAvailableError, CourseNotFoundError
     """
     course = _retrieve_course(course_key)
-    try:
-        requested_course_structure = models.CourseStructure.objects.get(course_id=course.id)
-        return serializers.CourseStructureSerializer(requested_course_structure.structure).data
-    except models.CourseStructure.DoesNotExist:
-        # If we don't have data stored, generate it and return an error.
-        tasks.update_course_structure.delay(unicode(course_key))
-        raise CourseStructureNotAvailableError
+
+    modified_timestamp = models.CourseStructure.objects.filter(course_id=course_key).values('modified')
+    if modified_timestamp.exists():
+        cache_key = 'openedx.content.course_structures.api.v0.api.course_structure.{}.{}.{}'.format(
+            course_key, modified_timestamp[0]['modified'], '_'.join(block_types or [])
+        )
+        data = cache.get(cache_key)  # pylint: disable=maybe-no-member
```

Comment:
can we add test around the caching to verify that we do not hit database once we set cache.